crates/typst-pdf/src/page.rs

```diff
@@ -711,6 +724,129 @@ fn write_group(ctx: &mut PageContext, pos: Point, group: &GroupItem) {
+    // we can render it directly
+    let sbix = match text.font.ttf().tables().sbix {
+        None => {
+            write_normal_text(ctx, pos, text);
+            return;
+        }
+        Some(sbix) => sbix,
+    };
+
+    let strike = sbix.best_strike(text.size.to_f32() as u16).unwrap();
+    let is_emoji = |g: &Glyph| strike.get(GlyphId(g.id)).is_some();
+    let emoji_count = text.glyphs.iter().filter(|g| is_emoji(*g)).count();
+
+    if emoji_count == text.glyphs.len() {
+        write_emojis(ctx, pos, text);
+    } else if emoji_count == 0 {
+        write_text(ctx, pos, text)
+    } else {
+        // Otherwise we need to split it in smaller text runs
+        let mut offset = 0;
+        let mut position_in_run = Em::zero();
+        while offset < text.glyphs.len() {
+            // Start a new text run where the last one ended
+            let start = offset;
+            // Determine if this is an emoji-only or a text-only run
+            let in_emoji_group = is_emoji(&text.glyphs[start]);
+            // Determine the index of the last glyph of the run
+            let end = start
+                + text.glyphs[start..]
+                    .iter()
+                    .position(|g| is_emoji(&g) != in_emoji_group)
+                    .unwrap_or(text.glyphs.len() - start);
+
+            // Build a sub text-run
+            let text_start = text.glyphs[start].range().start;
+            let text_end = text.glyphs[end - 1].range().end;
+
+            let new_text_item = TextItem {
+                text: EcoString::from(&text.text[text_start..text_end]),
+                // Glyphs ranges need to be re-mapped to the new text
+                glyphs: text.glyphs[start..end]
+                    .iter()
+                    .map(|g| Glyph {
+                        range: (g.range.start - text_start as u16)
+                            ..(g.range.end - text_start as u16),
+                        ..*g
+                    })
+                    .collect(),
+                font: text.font.clone(),
+                fill: text.fill.clone(),
+                stroke: text.stroke.clone(),
+                ..*text
+            };
+
+            // Adjust the position of the run on the line
+            let pos = pos + Point::new(position_in_run.at(text.size), Abs::zero());
+            // Actually write the text or emojis
+            if in_emoji_group {
+                write_emojis(ctx, pos, &new_text_item);
+            } else {
+                write_normal_text(ctx, pos, &new_text_item);
+            }
+
+            position_in_run += new_text_item.glyphs.iter().map(|g| g.x_advance).sum();
+            offset = end;
+        }
+    }
+}
+
+// Encodes a text run made only of emojis into the content stream
+fn write_emojis(ctx: &mut PageContext, pos: Point, text: &TextItem) {
+    let x = pos.x.to_f32();
+    let y = pos.y.to_f32();
+
+    if let Some(sbix) = text.font.ttf().tables().sbix {
```

Comment:
Is there a reason to query the table directly instead of using the more high-level [`glyph_raster_image`](https://docs.rs/ttf-parser/latest/ttf_parser/struct.Face.html#method.glyph_raster_image) method?

Reply:
Not really, I just didn't notice it existed when I started working on this. I will change that, thanks :)

Reply:
No worries!